rasa/core/policies/rule_policy.py

```diff
@@ -270,6 +277,137 @@ def _check_rule_restriction(
                 f"Please use stories for these cases."
             )
 
+    @staticmethod
+    def _check_slots_fingerprint(
+        fingerprint: Dict[Text, List[Text]], state: State, featurized_slots: Set[Text]
+    ) -> Set[Text]:
+        # leave only featurized slots
+        expected_slots = set(fingerprint.get(ACTION_FINGERPRINT_SLOTS)).intersection(
+            featurized_slots
+        )
+        current_slots = set(state.get(SLOTS, {}).keys())
+        if expected_slots == current_slots:
+            # all expected slots are satisfied
+            return set()
+
+        return expected_slots
+
+    @staticmethod
+    def _check_active_loops_fingerprint(
+        fingerprint: Dict[Text, List[Text]], state: State
+    ) -> Set[Text]:
+        expected_active_loops = set(fingerprint.get(ACTION_FINGERPRINT_ACTIVE_LOOPS))
+        # we don't use tracker.active_loop_name
+        # because we need to keep should_not_be_set
+        current_active_loop = state.get(ACTIVE_LOOP, {}).get(LOOP_NAME)
+        if current_active_loop in expected_active_loops:
+            # one of expected active loops is set
+            return set()
+
+        return expected_active_loops
+
+    @staticmethod
+    def _error_messages_from_fingerprints(
+        action_name: Text,
+        fingerprint_slots: Set[Text],
+        fingerprint_active_loops: Set[Text],
+        rule_name: Text,
+    ) -> List[Text]:
+        error_messages = []
+        if action_name and fingerprint_slots:
+            error_messages.append(
+                f"- the action '{action_name}' in rule '{rule_name}' does not set "
+                f"slots, that it sets in other rules: "
+                f"'{', '.join(fingerprint_slots)}'. Please update the rule with "
+                f"an appropriate slot or if it is the last action add "
+                f"`wait_for_user_input: false` after this action."
+            )
+        if action_name and fingerprint_active_loops:
+            # substitute `SHOULD_NOT_BE_SET` with `null` so that users
+            # know what to put in their rules
+            fingerprint_active_loops = set(
+                "null" if active_loop == SHOULD_NOT_BE_SET else active_loop
+                for active_loop in fingerprint_active_loops
+            )
+            # add action_name to active loop so that users
+            # know what to put in their rules
+            fingerprint_active_loops.add(action_name)
+
+            error_messages.append(
```

Comment:
@b-quachtran Could you please also double check the error messages for phrasing. Thanks 🙌

Reply:
cc @erohmensing as vova tagged you for the docs